agent/agent.go

```diff
@@ -588,79 +578,113 @@ func (r *agent) getStatusFrom(ctx context.Context, member serf.Member, respc cha
 	}
 }
 
-// statusResponse describes a status response from a background process that obtains
-// health status on the specified serf node.
-type statusResponse struct {
-	*pb.NodeStatus
-	member serf.Member
-	err    error
-}
+// timelineUpdateLoop periodically updates the status timeline by comparing
+// changes in local status and merging a timeline from another member.
+func (r *agent) timelineUpdateLoop() {
+	ticker := r.timelineClock.NewTicker(statusUpdateTimeout)
+	defer ticker.Stop()
 
-// recentStatus returns the last known cluster status.
-func (r *agent) recentStatus() (status *pb.SystemStatus, err error) {
-	status, err = r.cache.RecentStatus()
-	if err == nil && status == nil {
-		status = pb.EmptyStatus()
+	for range ticker.Chan() {
+		if r.isDone() {
+			return
+		}
+
+		if err := r.updateTimeline(); err != nil {
+			log.WithError(err).Warnf("Error updating timeline.")
+		}
 	}
-	return status, trace.Wrap(err)
 }
 
-// recentLocalStatus returns the last known local node status.
-func (r *agent) recentLocalStatus() *pb.NodeStatus {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	return r.localStatus
-}
+// updateTimeline updates the timeline with any new status changes.
+func (r *agent) updateTimeline() error {
+	ctx, cancel := context.WithTimeout(context.Background(), r.statusQueryReplyTimeout)
+	defer cancel()
+
+	// Update local status changes
+	if status := r.LocalStatus(); status != nil {
+		if err := r.Timeline.RecordStatus(ctx, status); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Collect timeline from another cluster member
+	timeline, err := r.collectTimeline(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
 
-func toMemberStatus(status string) pb.MemberStatus_Type {
-	switch MemberStatus(status) {
-	case MemberAlive:
-		return pb.MemberStatus_Alive
-	case MemberLeaving:
-		return pb.MemberStatus_Leaving
-	case MemberLeft:
-		return pb.MemberStatus_Left
-	case MemberFailed:
-		return pb.MemberStatus_Failed
-	}
-	return pb.MemberStatus_None
+	// Merge timelines
+	if timeline != nil {
+		if err := r.Timeline.RecordTimeline(ctx, timeline); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
 }
 
-// unknownNodeStatus creates an `unknown` node status for a node specified with member.
-func unknownNodeStatus(member *serf.Member) *pb.NodeStatus {
-	return &pb.NodeStatus{
-		Name:         member.Name,
-		Status:       pb.NodeStatus_Unknown,
-		MemberStatus: statusFromMember(member),
+// collectTimeline collects timeline from a random member.
```

Comment:
To be honest, I'm not a big fan of using randomness here. I'm not sure using random will even qualify as an eventually consistent algorithm.

The way I see it, we have two possible approaches to synchronizing the timeline we have to pick between: "pull" vs "push".

The "pull" approach is what you implemented here but instead of using random, I think that at the very least it should use a some sort of round-robin so it is guaranteed that we eventually grab the timeline from all nodes in a predictable fashion. The upside of this approach is that the node that does it (I'm assuming it only runs on masters) has full control over how quickly it iterates and can balance the workload. The downside is that on large clusters it can take a considerable amount of time to collect timeline from all nodes and when processing the last node, the status of the first one may have already changed (I guess this is what constitutes the "eventual" consistency property).

The "push" approach is what I mentioned in the comment above and what I believe we discussed in the design doc. In this scenario each node sends events to all master nodes as soon as they happen (can use some batching). With this approach the timeline on masters will always be consistent but the downside is that all nodes send their stuff to masters. However these events don't happen often so I don't think performance/scalability would be a concern.

We should also factor in various implementation details as well obviously, such as that if a node failed to send events, it should do retries, etc.

I'm personally leaning towards the "push" approach, just because it will make the timeline a much more useful real-time troubleshooting tool, but let me know what you think. cc @a-palchikov @knisbet as well.

Reply:
I guess you're right that using random would disqualify this algorithm from being eventually consistent. Will change to a round robin approach.

I agree that the "push" approach will be more efficient. I am working on implementing this feature incrementally, so I decided to go with the "pull" approach first because it was easier to implement. After testing for correctness I was planning on optimizing.